telemetry-common/src/main/java/com/sequenceiq/cloudbreak/telemetry/diagnostics/DiagnosticsOperationsService.java

```diff
@@ -0,0 +1,160 @@
+                m.getGatewayConfigs(), m.getNodes(), p.toMap(), m.getExitCriteriaModel()));
+    }
+
+    public void upload(Long stackId, DiagnosticParameters parameters) throws CloudbreakOrchestratorFailedException {
+        executeDiagnosticsOperation(stackId, parameters, "upload", (m, p) -> telemetryOrchestrator.uploadCollectedDiagnostics(
+                m.getGatewayConfigs(), m.getNodes(), p.toMap(), m.getExitCriteriaModel()));
+    }
+
+    public void cleanup(Long stackId, DiagnosticParameters parameters) throws CloudbreakOrchestratorFailedException {
+        executeDiagnosticsOperation(stackId, parameters, "cleanup", (m, p) -> telemetryOrchestrator.cleanupCollectedDiagnostics(
+                m.getGatewayConfigs(), m.getNodes(), p.toMap(), m.getExitCriteriaModel()));
+    }
+
+    public void vmPreflightCheck(Long stackId, DiagnosticParameters parameters) throws CloudbreakOrchestratorFailedException {
+        executeDiagnosticsOperation(stackId, parameters, "vm preflight check", (m, p) -> telemetryOrchestrator.preFlightDiagnosticsCheck(
+                m.getGatewayConfigs(), m.getNodes(), p.toMap(), m.getExitCriteriaModel()));
+    }
+
+    public void applyUnresponsiveHosts(Long stackId, DiagnosticParameters parameters)
+            throws CloudbreakOrchestratorFailedException {
+        OrchestratorMetadata orchestratorMetadata = orchestratorMetadataProvider.getOrchestratorMetadata(stackId);
+        OrchestratorMetadataFilter filter = OrchestratorMetadataFilter.Builder.newBuilder()
+                .includeHosts(parameters.getHosts())
+                .includeHostGroups(parameters.getHostGroups())
+                .exlcudeHosts(parameters.getExcludeHosts())
+                .build();
+        Set<Node> allUnresponsiveNodes = telemetryOrchestrator.collectUnresponsiveNodes(
+                orchestratorMetadata.getGatewayConfigs(), orchestratorMetadata.getNodes(), orchestratorMetadata.getExitCriteriaModel());
+        Set<String> unresponsiveHosts = allUnresponsiveNodes.stream()
+                .filter(filter::apply)
+                .map(Node::getHostname).collect(Collectors.toSet());
+        LOGGER.debug("Diagnostics collection salt validation operation has been started. parameters: '{}' stack id: {} ", parameters.toMap(), stackId);
+        if (CollectionUtils.isNotEmpty(unresponsiveHosts)) {
+            if (parameters.getSkipUnresponsiveHosts()) {
+                parameters.getExcludeHosts().addAll(unresponsiveHosts);
+                LOGGER.debug("Diagnostics parameters has been updated with excluded hosts. parameters: '{}', stack id: {}", parameters.toMap(), stackId);
+            } else {
+                throw new CloudbreakOrchestratorFailedException(
+                        String.format("Some of the hosts are unresponsive, check the states of salt-minions on the following nodes: %s",
+                                StringUtils.join(unresponsiveHosts, ',')));
+            }
+        }
+    }
```

Comment:
I would split this and change the method signature:
```
    public DiagnosticParameters applyUnresponsiveHosts(Long stackId, DiagnosticParameters parameters)
            throws CloudbreakOrchestratorFailedException {
        Set<String> unresponsiveHosts = collectUnresponsiveHosts(stackId, parameters);
        LOGGER.debug("Diagnostics collection salt validation operation has been started. parameters: '{}' stack id: {} ", parameters.toMap(), stackId);
        if (CollectionUtils.isNotEmpty(unresponsiveHosts)) {
            if (parameters.getSkipUnresponsiveHosts()) {
                parameters.getExcludeHosts().addAll(unresponsiveHosts);
                LOGGER.debug("Diagnostics parameters has been updated with excluded hosts. parameters: '{}', stack id: {}", parameters.toMap(), stackId);
                return parameters;
            } else {
                throw new CloudbreakOrchestratorFailedException(
                        String.format("Some of the hosts are unresponsive, check the states of salt-minions on the following nodes: %s",
                                StringUtils.join(unresponsiveHosts, ',')));
            }
        } else {
            //LOG
            return parameters;
        }
    }

    private Set<String> collectUnresponsiveHosts(Long stackId, DiagnosticParameters parameters) throws CloudbreakOrchestratorFailedException {
        Set<Node> allUnresponsiveNodes = collectUnresponsiveNodes(stackId);
        OrchestratorMetadataFilter filter = createMetadataFilter(parameters);
        return allUnresponsiveNodes.stream()
                .filter(filter::apply)
                .map(Node::getHostname).collect(Collectors.toSet());
    }

    private Set<Node> collectUnresponsiveNodes(Long stackId) throws CloudbreakOrchestratorFailedException {
        OrchestratorMetadata orchestratorMetadata = orchestratorMetadataProvider.getOrchestratorMetadata(stackId);
        return telemetryOrchestrator.collectUnresponsiveNodes(
                orchestratorMetadata.getGatewayConfigs(), orchestratorMetadata.getNodes(), orchestratorMetadata.getExitCriteriaModel());
    }

    private OrchestratorMetadataFilter createMetadataFilter(DiagnosticParameters parameters) {
        return OrchestratorMetadataFilter.Builder.newBuilder()
                .includeHosts(parameters.getHosts())
                .includeHostGroups(parameters.getHostGroups())
                .exlcudeHosts(parameters.getExcludeHosts())
                .build();
    }
```
- creating `createMetadataFilter` replaces duplicate code in a separate method
- I have reordered stuff to be created right before it's used, in this way you can easily group things together
- on the caller validation handler side you should have the code modified like this:
```
DiagnosticParameters parameters = diagnosticsOperationsService.applyUnresponsiveHosts(resourceId, data.getParameters());
```
so it's obvious the parameters are modified. maybe rename this apply method